src/types.py

```diff
@@ -273,13 +374,88 @@ def of_size(size: int) -> "Type":
     def bool() -> "Type":
         return Type.intish()
 
+    @staticmethod
+    def void() -> "Type":
+        return Type(kind=Type.K_VOID, size=0)
+
+
+@attr.s(eq=False)
+class FunctionParam:
+    type: Type = attr.ib(factory=Type.any)
+    name: str = attr.ib(default="")
+
+
+@attr.s(eq=False)
+class FunctionSignature:
+    return_type: Type = attr.ib(factory=Type.any)
+    params: List[FunctionParam] = attr.ib(factory=list)
+    params_known: bool = attr.ib(default=False)
+    is_variadic: bool = attr.ib(default=False)
+
+    def unify(self, other: "FunctionSignature") -> bool:
+        if self.is_variadic != other.is_variadic:
+            return False
+
+        # Try to unify *all* ret/param types, without returning early
+        can_unify = True
+        if self.params_known and other.params_known:
+            if len(self.params) != len(other.params):
+                return False
+            for x, y in zip(self.params, other.params):
+                can_unify &= x.type.unify(y.type)
+        can_unify &= self.return_type.unify(other.return_type)
+
+        if can_unify:
+            if not self.params_known:
+                self.params = other.params
+                self.params_known = other.params_known
+            elif not other.params_known:
+                other.params = self.params
+                other.params_known = self.params_known
```

Comment:
If neither params_known is true, is it worth trying to unify up to the min of the two param lengths, and setting both `params` equal to the longer one?